sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/CacheAfterWriteStream.cs

```diff
@@ -0,0 +1,195 @@
+    /// properties (e.g. Blob's eTag) populated by the azure-sdk-for-net storage extension.
+    /// </summary>
+    internal class CacheAfterWriteStream : Stream, ICacheAfterWriteStream
+    {
+        // Metadata of where the object is present in shared memory.
+        // When inserting the object into the cache, this will be used.
+        private readonly SharedMemoryMetadata _sharedMemoryMeta;
+
+        // The cache must not be disposed.
+#pragma warning disable CA2213 // Disposable fields should be disposed
+        private readonly IFunctionDataCache _functionDataCache;
+#pragma warning restore CA2213 // Disposable fields should be disposed
+
+        // Blob representing the object that will be written into the cache.
+        private readonly BlobWithContainer<BlobBaseClient> _blob;
+
+        // Inner Stream pointing to an object in remote storage (e.g. Blob).
+        private readonly Stream _inner;
+
+        public CacheAfterWriteStream(SharedMemoryMetadata sharedMemoryMeta, IFunctionDataCache functionDataCache, BlobWithContainer<BlobBaseClient> blob, Stream innerStream)
+        {
+            _sharedMemoryMeta = sharedMemoryMeta;
+            _functionDataCache = functionDataCache;
+            _blob = blob;
+            _inner = innerStream;
+        }
+
+        public async Task<bool> TryPutToFunctionDataCacheAsync(bool isDeleteOnFailure)
+        {
+            if (_sharedMemoryMeta == null)
+            {
+                return false;
+            }
+
+            BlobProperties properties = await _blob.BlobClient.FetchPropertiesOrNullIfNotExistAsync().ConfigureAwait(false);
+            return TryPutToFunctionDataCacheCore(properties, isDeleteOnFailure);
+        }
+
+        /// <summary>
+        /// Put the object into the cache by generating a key for the object based on the blob's properties.
+        /// </summary>
+        /// <param name="properties">Properties of the blob corresponding to the object being written.</param>
+        /// <param name="isDeleteOnFailure">If True, in the case where the cache is unable to insert this object, the local resources pointed to by the Stream (which were to be cached) will be deleted.</param>
+        /// <returns>True if the object was written to the <see cref="IFunctionDataCache"/>, false otherwise.</returns>
+        private bool TryPutToFunctionDataCacheCore(BlobProperties properties, bool isDeleteOnFailure)
+        {
+            string eTag = properties.ETag.ToString();
+            string id = _blob.BlobClient.Uri.ToString();
+            FunctionDataCacheKey cacheKey = new FunctionDataCacheKey(id, eTag);
+            return _functionDataCache.TryPut(cacheKey, _sharedMemoryMeta, isIncrementActiveReference: false, isDeleteOnFailure);
+        }
+
+        public override bool CanRead
+        {
+            get { return _inner.CanRead; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return _inner.CanSeek; }
+        }
```

Comment:
You can leverage https://github.com/Azure/azure-sdk-for-net/blob/17a6cbdeff99e7e881e27cc30da68e649f36966f/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/DelegatingStream.cs#L11 to simplify this class. Similarly to https://github.com/Azure/azure-sdk-for-net/blob/17a6cbdeff99e7e881e27cc30da68e649f36966f/sdk/storage/Microsoft.Azure.WebJobs.Extensions.Storage.Blobs/src/Bindings/NotifyingBlobStream.cs#L10